adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,355 @@
+package richaudience
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/metrics"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type richaudienceRequest struct {
+	ID     string             `json:"id,omitempty"`
+	Imp    []openrtb2.Imp     `json:"imp,omitempty"`
+	User   richaudienceUser   `json:"user,omitempty"`
+	Device richaudienceDevice `json:"device,omitempty"`
+	Site   richaudienceSite   `json:"site,omitempty"`
+	Test   int8               `json:"test,omitempty"`
+}
+
+type richaudienceUser struct {
+	BuyerUID string              `json:"buyeruid,omitempty"`
+	Ext      richaudienceUserExt `json:"ext,omitempty"`
+}
+
+type richaudienceUserExt struct {
+	Eids    []openrtb_ext.ExtUserEid `json:"eids,omitempty"`
+	Consent string                   `json:"consent,omitempty"`
+}
+
+type richaudienceDevice struct {
+	//GEO only Java, IFA
+	IP  string `json:"ip,omitempty"`
+	Lmt int8   `json:"lmt,omitempty"`
+	DNT int8   `json:"dnt,omitempty"`
+	UA  string `json:"ua,omitempty"`
+}
+type richaudienceSite struct {
+	//Site: Cat IAB
+	Domain string `json:"domain,omitempty"`
+	Page   string `json:"page,omitempty"`
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+
+	if config.Endpoint != "" && strings.Contains(config.Endpoint, "richaudience") {
+		bidder := &adapter{
+			endpoint: config.Endpoint,
+		}
+
+		return bidder, nil
+	}
+
+	return nil, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	richaudienceRequest := richaudienceRequest{}
+
+	raiHeaders := http.Header{}
+
+	raiErrors := make([]error, 0, len(request.Imp))
```

Comment:
I like the comprehensive error checking you have, but there is no need for this variable. The slice will only have at most one error since you return it immediately. Please remove this variable and change the return methods to something like, just like you do in `MakeBids`:

```
return nil, []error{&errortypes.BadInput{Message: err.Error()})}
```

Reply:
in next commit i have solved